### PR TITLE
add node network connected subnet route to UDN VRF table

### PR DESCRIPTION
**Pull request by rameshsahoo11**
## 📑 Description

Using OVN local gateway mode, the NATted traffic from UDN pods destined to nodes on the same L2 subnet was being routed through the default gateway instead of being forwarded directly via br-ex/breth0. This happened because the VRF table doesn not have a connected subnet route for the node network.

Add a scope-link route (e.g. 192.168.x.0/24 dev br-ex/breth0 scope link src 192.168.x.5) to the UDN VRF table so the kernel can resolve destinations on the local subnet via direct L2/ARP forwarding without an unnecessary hop through the gateway.

Fixes: https://issues.redhat.com/browse/OCPBUGS-78351

## Additional Information for reviewers

1. gateway_udn.go — The fix (3 changes)

  a) New struct field — Added nodeNetAddrs []*net.IPNet to UserDefinedNetworkGateway to store the gateway bridge (br-ex) IP addresses with their subnet masks.

  b) Populate in constructor — In NewUserDefinedNetworkGateway(), fetch the bridge IPs from openflowManager.defaultBridge.GetIPs() when available. Guarded by a nil check since
  openflowManager is nil in DPUHost mode.

  c) New route in computeRoutesForUDN() — After the default gateway route, added a loop that creates a scope-link connected subnet route for each bridge IP:

```go
  for _, nodeNetAddr := range udng.nodeNetAddrs {
      nodeSubnet := nodeNetAddr.IP.Mask(nodeNetAddr.Mask)
      retVal = append(retVal, netlink.Route{
          LinkIndex: udng.gwInterfaceIndex,
          Dst:       &net.IPNet{IP: nodeSubnet, Mask: nodeNetAddr.Mask},
          Src:       nodeNetAddr.IP,
          Table:     udng.vrfTableId,
          Scope:     netlink.SCOPE_LINK,
      })
  }
```

  This produces: 192.168.x.0/24 dev br-ex scope link src 192.168.x.5 in the VRF table.

  2. bridgeconfig/bridgeconfig_testutil.go — Test helper

  Added a SetIPs() method on BridgeConfiguration so tests can configure bridge IPs. Follows the same mutex pattern as the existing GetIPs().

  3. gateway_udn_test.go — Test coverage (2 changes)

  a) Updated getDummyOpenflowManager — Changed signature to variadic bridgeIPs ...*net.IPNet so tests can optionally pass bridge IPs. All 13 existing callers continue to work unchanged.

  b) New test case — "should add node network connected subnet routes to VRF table" verifies both IPv4 (192.168.1.0/24) and IPv6 (fc00:f853:ccd:e793::/64) connected subnet routes are present
  with correct Dst, Src, LinkIndex, Table, and Scope. Uses order-independent matching by looping over routes.

## ✅ Checks
Local Unit test succeeded and the following are the  Unit Test Results

```bash
  Test Suite: pkg/node/... (gateway_udn, bridgeconfig, vrfmanager, and related packages)
  Branch: udn-vrf-node-subnet-route

  Related computeRoutesForUDN tests — all PASSED:
    ✔ should compute correct routes for a user defined network
    ✔ should have default route when network is advertised on default VRF
    ✔ should omit default route when network is advertised on any other vrf than default
    ✔ should create a route import VRF in DPU mode
    ✔ should add node network connected subnet routes to VRF table   <----------  NEW TEST

  All suites:
    ✔ pkg/node             — 210 Passed | 0 Failed
    ✔ pkg/node/bridgeconfig —   2 Passed | 0 Failed
    ✔ pkg/node/controllers  —  43 Passed | 0 Failed
    ✔ pkg/node/controllers/egressip — 14 Passed | 0 Failed
    ✔ pkg/node/managementport — 51 Passed | 0 Failed
    ✔ pkg/node/netlinkdevicemanager — 206 Passed | 0 Failed
    ✔ pkg/node/routemanager — 20 Passed | 0 Failed
    ✔ pkg/node/util         —  4 Passed | 0 Failed
    ✔ pkg/node/vrfmanager   —  8 Passed | 0 Failed

  Total: 0 Failed | 0 Regressions
```

## How to verify it
- Local Unit test succeeded.
- In an OVN  kind cluster using the modified code in local gateway mode, validated that UDN traffic destined to node subnet, leaves via direct L2/ARP forwarding without an unnecessary hop through the gateway.
- Validated that scope-link route existed in the UDN vrf route table on node. 

### Before Fix

```bash
➜ ./kind.sh -mne -nse -gm local -wk 2 

➜ oc get nodes -o wide
NAME                STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION           CONTAINER-RUNTIME
ovn-control-plane   Ready    control-plane   13m   v1.35.0   172.18.0.4    <none>        Debian GNU/Linux 12 (bookworm)   7.0.10-101.fc43.x86_64   containerd://2.2.0
ovn-worker          Ready    worker          13m   v1.35.0   172.18.0.3    <none>        Debian GNU/Linux 12 (bookworm)   7.0.10-101.fc43.x86_64   containerd://2.2.0
ovn-worker2         Ready    worker          13m   v1.35.0   172.18.0.2    <none>        Debian GNU/Linux 12 (bookworm)   7.0.10-101.fc43.x86_64   containerd://2.2.0

➜ oc get po -o wide
NAME   READY   STATUS    RESTARTS   AGE     IP           NODE          NOMINATED NODE   READINESS GATES
pod1   1/1     Running   0          2m48s   10.244.2.3   ovn-worker    <none>           <none>
pod2   1/1     Running   0          2m41s   10.244.1.6   ovn-worker2   <none>           <none>

[root@pod1 /]# curl -v telnet://172.18.0.2:8888
*   Trying 172.18.0.2:8888...
* Connected to 172.18.0.2 (172.18.0.2) port 8888

root@ovn-worker:~# ip a s breth0
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 46:fe:5b:3d:21:dc brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 brd 172.18.255.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet 169.254.0.2/17 brd 169.254.127.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet6 fc00:f853:ccd:e793::3/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::44fe:5bff:fe3d:21dc/64 scope link 
       valid_lft forever preferred_lft forever

root@ovn-worker:~# ip vrf
Name              Table
-----------------------
mp1-udn-vrf       1009

root@ovn-worker:~# ip route show table 1009
default via 172.18.0.1 dev breth0 mtu 1400 
unreachable default metric 4278198272 
10.96.0.0/16 via 169.254.0.4 dev breth0 mtu 1400 
10.200.0.0/16 dev ovn-k8s-mp1 proto kernel scope link src 10.200.0.2 
local 10.200.0.2 dev ovn-k8s-mp1 proto kernel scope host src 10.200.0.2 
broadcast 10.200.255.255 dev ovn-k8s-mp1 proto kernel scope link src 10.200.0.2 
169.254.0.3 via 10.200.0.1 dev ovn-k8s-mp1 
169.254.0.12 dev ovn-k8s-mp1 mtu 1400 

// In the above 1009 route table there is no scope-link route for 172.18.0.0/0 network 
// 8a:af:c0:35:d2:c7 is the MAC address of node's default gateway 

// UDN POD traffic destined to 172.18.0.2.8888 is forwarded directly to the gateway MAC 8a:af:c0:35:d2:c7
root@ovn-worker:~# tcpdump -i eth0 -nnne port 8888
16:33:00.749256 46:fe:5b:3d:21:dc > 8a:af:c0:35:d2:c7, ethertype IPv4 (0x0800), length 74: 172.18.0.3.34622 > 172.18.0.2.8888: Flags [S], seq 1496132001, win 65280, options [mss 1360,sackOK,TS val 3344983273 ecr 0,nop,wscale 10], length 0
16:33:00.749787 8a:af:c0:35:d2:c7 > 46:fe:5b:3d:21:dc, ethertype IPv4 (0x0800), length 74: 172.18.0.2.8888 > 172.18.0.3.34622: Flags [S.], seq 1953367920, ack 1496132002, win 65160, options [mss 1460,sackOK,TS val 3938482833 ecr 3344983273,nop,wscale 10], length 0

// On node 2, UDN POD traffic is received via default gateway 172.18.0.1/8a:af:c0:35:d2:c7 
root@ovn-worker2:~# ip a s breth0
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 86:a5:6b:84:c8:5d brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.2/16 brd 172.18.255.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet 169.254.0.2/17 brd 169.254.127.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet6 fc00:f853:ccd:e793::2/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::84a5:6bff:fe84:c85d/64 scope link 
       valid_lft forever preferred_lft forever

root@ovn-worker2:~# tcpdump -i breth0 -nnne port 8888 
16:33:00.749574 8a:af:c0:35:d2:c7 > 86:a5:6b:84:c8:5d, ethertype IPv4 (0x0800), length 74: 172.18.0.1.34622 > 172.18.0.2.8888: Flags [S], seq 1496132001, win 65280, options [mss 1360,sackOK,TS val 3344983273 ecr 0,nop,wscale 10], length 0
16:33:00.749627 86:a5:6b:84:c8:5d > 8a:af:c0:35:d2:c7, ethertype IPv4 (0x0800), length 74: 172.18.0.2.8888 > 172.18.0.1.34622: Flags [S.], seq 1953367920, ack 1496132002, win 65160, options [mss 1460,sackOK,TS val 3938482833 ecr 3344983273,nop,wscale 10], length 0

// gateway mac 8a:af:c0:35:d2:c7
5: br-233e6d1b2036: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default 
    link/ether 8a:af:c0:35:d2:c7 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 brd 172.18.255.255 scope global br-233e6d1b2036
       valid_lft forever preferred_lft forever
    inet6 fc00:f853:ccd:e793::1/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::88af:c0ff:fe35:d2c7/64 scope link proto kernel_ll 
       valid_lft forever preferred_lft forever
```

### After Fix

```bash
➜ ./kind.sh -mne -nse -gm local -wk 2 

➜ oc get nodes -o wide
NAME                STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                      
ovn-control-plane   Ready    control-plane   34m   v1.35.0   172.18.0.2    <none>        Debian GNU/Linux 12 (bookworm)
ovn-worker          Ready    worker          34m   v1.35.0   172.18.0.4    <none>        Debian GNU/Linux 12 (bookworm)
ovn-worker2         Ready    worker          33m   v1.35.0   172.18.0.3    <none>        Debian GNU/Linux 12 (bookworm)

➜ oc get po -o wide
NAME   READY   STATUS    RESTARTS   AGE   IP           NODE          NOMINATED NODE   READINESS GATES
pod1   1/1     Running   0          35s   10.244.1.3   ovn-worker    <none>           <none>
pod2   1/1     Running   0          30s   10.244.2.6   ovn-worker2   <none>           <none>

[root@pod1 /]# curl -v telnet://172.18.0.2:8888
*   Trying 172.18.0.2:8888...
* Connected to 172.18.0.2 (172.18.0.2) port 8888

root@ovn-worker:~# ip a s breth0
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether fa:a6:44:ec:1d:f6 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.4/16 brd 172.18.255.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet 169.254.0.2/17 brd 169.254.127.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet6 fc00:f853:ccd:e793::4/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::f8a6:44ff:feec:1df6/64 scope link 
       valid_lft forever preferred_lft forever

root@ovn-worker:~# ip vrf
Name              Table
-----------------------
mp1-udn-vrf       1009

root@ovn-worker:~# ip route show table 1009
default via 172.18.0.1 dev breth0 mtu 1400 
unreachable default metric 4278198272 
10.96.0.0/16 via 169.254.0.4 dev breth0 mtu 1400 
10.200.0.0/16 dev ovn-k8s-mp1 proto kernel scope link src 10.200.0.2 
local 10.200.0.2 dev ovn-k8s-mp1 proto kernel scope host src 10.200.0.2 
broadcast 10.200.255.255 dev ovn-k8s-mp1 proto kernel scope link src 10.200.0.2 
169.254.0.3 via 10.200.0.1 dev ovn-k8s-mp1 
169.254.0.12 dev ovn-k8s-mp1 mtu 1400 
172.18.0.0/16 dev breth0 scope link src 172.18.0.4  <<<<<  Connected subnet route

// N/W packets are directly forwarded to destination system 
root@ovn-worker:~# tcpdump -i eth0 -nnne port 8888
17:56:04.384520 fa:a6:44:ec:1d:f6 > 8a:d5:fd:43:ec:e6, ethertype IPv4 (0x0800), length 74: 172.18.0.4.40238 > 172.18.0.3.8888: Flags [S], seq 3810489716, win 65280, options [mss 1360,sackOK,TS val 858594941 ecr 0,nop,wscale 10], length 0
17:56:04.384682 8a:d5:fd:43:ec:e6 > fa:a6:44:ec:1d:f6, ethertype IPv4 (0x0800), length 74: 172.18.0.3.8888 > 172.18.0.4.40238: Flags [S.], seq 4222349067, ack 3810489717, win 65160, options [mss 1460,sackOK,TS val 1218116599 ecr 858594941,nop,wscale 10], length 0

// Node2 receives traffic from node1's MAC and IP directly. 
root@ovn-worker2:~# ip a s breth0
6: breth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether 8a:d5:fd:43:ec:e6 brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.3/16 brd 172.18.255.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet 169.254.0.2/17 brd 169.254.127.255 scope global breth0
       valid_lft forever preferred_lft forever
    inet6 fc00:f853:ccd:e793::3/64 scope global nodad 
       valid_lft forever preferred_lft forever
    inet6 fe80::88d5:fdff:fe43:ece6/64 scope link 
       valid_lft forever preferred_lft forever

root@ovn-worker2:~# tcpdump -i breth0 -nnne port 8888
17:56:04.384580 fa:a6:44:ec:1d:f6 > 8a:d5:fd:43:ec:e6, ethertype IPv4 (0x0800), length 74: 172.18.0.4.40238 > 172.18.0.3.8888: Flags [S], seq 3810489716, win 65280, options [mss 1360,sackOK,TS val 858594941 ecr 0,nop,wscale 10], length 0
17:56:04.384617 8a:d5:fd:43:ec:e6 > fa:a6:44:ec:1d:f6, ethertype IPv4 (0x0800), length 74: 172.18.0.3.8888 > 172.18.0.4.40238: Flags [S.], seq 4222349067, ack 3810489717, win 65160, options [mss 1460,sackOK,TS val 1218116599 ecr 858594941,nop,wscale 10], length 0
```

## Summary by CodeRabbit

* **New Features**
  * Gateway now caches bridge IPs and programs routes for user-defined networks, enabling direct Layer‑2 forwarding to nodes on the same subnets.

* **Tests**
  * Added a test helper to configure bridge IPs and expanded tests to verify VRF routing for node-connected subnets (IPv4 and IPv6).

### CODE REVIEW

**Comment by rameshsahoo11** [comment]
Latest modification after coderabbit review:

Updated the node network connected subnet route loop in computeRoutesForUDN() with two additional safeguards:

1. Defensive nil checks — Skip entries where nodeNetAddr, nodeNetAddr.IP, or nodeNetAddr.Mask is nil to prevent a panic:

```go
  if nodeNetAddr == nil || nodeNetAddr.IP == nil || nodeNetAddr.Mask == nil {
      continue
  }
```

2. IP family filtering — Only add routes for configured IP families, consistent with how getDefaultRoute() (line 614) and service CIDR routes (line 475) filter by address family:

```go
  isV6 := utilnet.IsIPv6(nodeNetAddr.IP)
  if (!isV6 && !config.IPv4Mode) || (isV6 && !config.IPv6Mode) {
      continue
  }
```

This ensures correct behavior across all configurations:

  - Dual-stack (IPv4Mode=true, IPv6Mode=true): both IPv4 and IPv6 routes added
  - IPv4-only (IPv4Mode=true, IPv6Mode=false): IPv4 route added, IPv6 skipped
  - IPv6-only (IPv4Mode=false, IPv6Mode=true): IPv4 skipped, IPv6 route added

The existing unit test "should add node network connected subnet routes to VRF table" runs with dual-stack enabled and validates both IPv4 and IPv6 routes are present.

Unit test result after the update:

```bash
  $ GINKGO_FOCUS="should add node network connected subnet routes to VRF table" PKGS=./pkg/node make test

  UserDefinedNetworkGateway should add node network connected subnet routes to VRF table
  • [0.135 seconds]

  Ran 1 of 210 Specs in 0.145 seconds
  SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 209 Skipped
  --- PASS: TestNodeSuite (0.19s)
PASS
```

File: go-controller/pkg/node/gateway_udn.go [hunk 1]
```diff
@@ -148,6 +151,12 @@ func NewUserDefinedNetworkGateway(netInfo util.NetInfo, node *corev1.Node, nodeL
 		return nil, fmt.Errorf("unable to get link for gateway interface %s, error: %v", intfName, err)
 	}
 
+	// Retrieve gateway bridge IPs for programming node network connected subnet routes in the VRF table
+	var nodeNetAddrs []*net.IPNet
+	if gw.openflowManager != nil {
+		nodeNetAddrs = gw.openflowManager.defaultBridge.GetIPs()
```
Comment (shreyasbe) [inline go-controller/pkg/node/gateway_udn.go:182]: If GetIPs() returns an empty:
Suggestion: Add a warning log here if len(nodeNetAddrs) == 0 to catch initialization race conditions early:
klog.V(4).Infof("No bridge IPs found for UDN gateway on node %s", node.Name)

File: go-controller/pkg/node/gateway_udn_test.go [hunk 2]
```diff
@@ -1785,6 +1790,68 @@ var _ = Describe("UserDefinedNetworkGateway", func() {
 		Expect(fexec.CalledMatchesExpected()).To(BeTrue(), fexec.ErrorDesc)
 	})
 
+	// Verifies that node network connected subnet routes (e.g. 192.168.1.0/24 dev br-ex scope link src 192.168.1.10)
```
Comment (shreyasbe) [inline go-controller/pkg/node/gateway_udn_test.go:2538]: Add some boundary conditions TC:
- Set config.IPv4Mode = false and verify the IPv6 route is added but the IPv4 route is correctly skipped (and vice versa).
- Inject a &net.IPNet{IP: nil, Mask: nil} into bridgeIPs to validate the defensive

File: go-controller/pkg/node/gateway_udn.go (hunk 1 above)
Comment (rameshsahoo11) [reply go-controller/pkg/node/gateway_udn.go:182]: Done. Added klog.V(4) warning when GetIPs() returns empty. Used netInfo.GetNetworkName() instead of node.Name since it's more actionable in this context — it directly identifies which
  network failed to get bridge IPs:

  if len(nodeNetAddrs) == 0 {
      klog.V(4).Infof("No bridge IPs found for UDN gateway %s, node network connected subnet routes will not be programmed", netInfo.GetNetworkName())
  }

File: go-controller/pkg/node/gateway_udn_test.go (hunk 2 above)
Comment (rameshsahoo11) [reply go-controller/pkg/node/gateway_udn_test.go:2538]: ● Added a single table-driven test covering all three boundary conditions in one compact block:

  for _, tc := range []struct {
      ipv4, ipv6     bool
      wantV4, wantV6 bool
      injectNil      bool
  }{
      {true, false, true, false, false},   // IPv4-only: IPv6 route skipped
      {false, true, false, true, false},   // IPv6-only: IPv4 route skipped
      {true, true, true, true, true},      // nil injection: no panic, both routes present
  } { ... }

  - IPv4Mode=false → IPv6 scope-link route present, IPv4 absent ✓
  - IPv6Mode=false → IPv4 scope-link route present, IPv6 absent ✓
  - nil entry in bridge IPs → skipped by the existing nil guard, valid routes still programmed ✓

**Review by shreyasbe — APPROVED** [review]
LGTM

**Review by shreyasbe — CHANGES_REQUESTED** [review]
Address CI Failure

**Review by shreyasbe — APPROVED** [review]
LGTM

**Comment by rameshsahoo11** [comment]
@jcaamano as discussed, kindly review this PR.

**Comment by jcaamano** [comment]
@rameshsahoo11 sorry this took so long to look at

I think we need to consider a couple of extra things here.

* Node addresses may change at runtime; so you might want to hook into `(c *addressManager) AddOnAddressesChangedHandler` to be notified when it happens and reconfigure the routes in that case.
* There is a new uplink feature that allows us to associate UDNs to different bridges/gateways. We would need to do something about that also unless there is a specific reason not to.

Let me know what you think. I will stay on top of the PR from now on.

**Comment by rameshsahoo11** [comment]
Thanks a lot @jcaamano  for the thoughtful review, your both points are well taken.

On the uplink feature (feedback 2): After digging into the datapath, the `connected-subnet` route only takes effect in local gateway mode. In shared gateway mode UDN pod egress stays within OVS and leaves via `breth0/br-ex` directly, so it never traverses the host VRF (the route is effectively a no-op there). 

Since the `uplink` feature is shared-gateway-only, I'd propose gating this route with `config.Gateway.Mode == config.GatewayModeLocal`, as I think `uplink` is therefore out of scope. This resolves the second feedback without a bridge-selection refactor, since a local-gateway UDN always uses the default bridge. Please let me know if you see a case I might be missing here.

On runtime address changes (first feedback): Good catch,  I'll handle this by hooking into `AddOnAddressesChangedHandler` so the route is re-reconciled when the default bridge's IPs change.

Would this approach work for you? Happy to adjust if you'd prefer handling uplink explicitly instead.
